index.js

```diff
@@ -546,14 +559,40 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<boolean>} Resolves to true once keyrings are persisted.
    */
-  persistAllKeyrings(password = this.password) {
-    if (typeof password !== 'string') {
+  persistAllKeyrings(password) {
+    if (password && typeof password !== 'string') {
       return Promise.reject(
         new Error('KeyringController - password is not a string'),
       );
     }
 
-    this.password = password;
+    // MV3: Since we also allow persisting without a password, we should require this.encryptedKey
+    if (password === undefined && this.encryptedKey === undefined) {
+      return Promise.reject(
+        new Error(
+          'KeyringController - a password or encryptedKey must exist to persist keyrings',
+        ),
+      );
+    }
+
+    let salt = null;
+    if (password) {
+      // MV3: If this is a migration or new password-driven login, we should
+      // create or rotate the salt
+      salt = this._generateSalt();
+
+      // MV3: Since there's a new salt, we need to generate a new encrypted key
+      // for use in the
+      this.encryptedKey = this._generateEncryptedKey(password, salt);
+    } else {
+      const encryptedVault = this.store.getState().vault;
+      if (!encryptedVault) {
+        throw new Error('Cannot unlock without a previous vault.');
+      }
+      // MV3: We can use an existing salt if one exists in the encrypted key
+      [, salt] = encryptedVault.split(VAULT_SEPARATOR);
```

Comment:
(could) In my (very limited) experience it pays off to hide the logic of extracting bits of string behind a tiny abstraction if you use it in more than one place. If we ever end up in a situation where we need to migrate from one separator to another or something, we're much better off having a test-covered abstraction where we can quickly introduce tests for the separator migration or any complications we have to add to the salt storage.

Reply:
Done!